File: java/src/Test1.java
import java.util.Scanner;
import java.util.Arrays;
import java.util.List;
import java.util.ArrayList;
import java.io.*;
import java.util.NoSuchElementException;

public class Test1 extends Test {
    public boolean runTest(File f, boolean verbose) {
	List<Integer> temperatures = TestHelpers.getIntegerListFromFile(f);
	List<Integer> actual = Problem1.itshotitscold(temperatures);
	File resultsFile = TestHelpers.getResultsFile(f);
	List<Integer> expected = TestHelpers.getIntegerListFromFile(resultsFile, 3);

	if (expected == null) {
	    // bad results file.
	    return false;
	}

	if (actual == null) {
	    if (verbose) {
		System.out.println("Failed: actual result is null");		
	    }
	    return false;
	} else if (actual.equals(expected)) {
	    return true;	    
	} else {
	    if (verbose) {
		System.out.println("Failed actual result and expected result do not match");
		printDiff(actual, expected);
	    }
	    return false;	    
	}
    }

    private static void printDiff(List<Integer> actual, List<Integer> expected) {
	if (actual.size() != expected.size()) {
	    String err = String.format("Actual length (%d) and expected length (%d) do not match",
				       actual.size(), expected.size());
	    System.out.println(err);
	    return;
	}

	for (int i = 0; i < actual.size(); i++) {
	    if (actual.get(i) != expected.get(i)) {
		System.out.println(String.format("  Actual and expected results do not match at index: %d", i));
		System.out.println(String.format("    actual @ index %d is %d", i, actual.get(i)));
		System.out.println(String.format("    expected @ index %d is %d", i, expected.get(i)));
		return;
	    }
	}
	return;
    }

    public static void main(String[] args) {
	System.exit(TestHelpers.doTests(args, new Test1()));
    }
}

File: cpp/src/Test1.cpp
#include "TestHelpers.h"
#include "Problem1.h"
#include <cstdio>
#include <iostream>
#include <fstream>
#include <vector>
#include <string>

class Test {
public:
static bool runTest(std::string f, bool verbose) {
    std::vector<int> temperatures = TestHelpers::getData(f);
    std::vector<int> actual = Problem1::itshotitscold(temperatures);
    std::vector<int> expected = TestHelpers::getResult(f);
    

    if (expected.size() == 0) {
        return false;
    }

    if (actual.size() == 0) {
        if (verbose) {
  	    std::cout << "Test: " << f << ": ";
            std::cout << "Failed: actual result is null\n";
        }  
	return false;
    } else if (actual == expected) {
        return true;
    } else {
      if (verbose) {
          std::cout << "Test: " << f << ": ";
          std::cout << "Failed actual result and expected result do not match\n";
          printDiff(actual, expected);
      }
      return false;
    }
}

static void printDiff(const std::vector<int>& actual, const std::vector<int>& expected) {
    if (actual.size() != expected.size()) {
        std::cout << "Actual length (" << actual.size() << ") and expected length (";
        std::cout << expected.size() << ") do not match";
        return;
    }

    for (int i = 0; i < actual.size(); ++i) {
        if (actual[i] != expected[i]) {
            std::cout << "  Actual and expected results do not match at index: " << i << std::endl;
            std::cout << "    actual @ index " << i << " is " << actual[i] << std::endl;
            std::cout << "    expected @ index " << i << " is " << expected[i] << std::endl;
            return;
        }
    }
    return;
}
};

int main (int argc, char* argv[]) {
    std::vector<std::string> files = TestHelpers::getFiles(argc, argv);
    if (files.size() == 0) {
        std::cout << "no test file found"; 
        exit(1);
    } 

    int count = 0;
    for (size_t i = 0; i < files.size(); ++i) {
        std::string f = files[i];
	std::cout << "Test " << f << ": ";	 	
        if (Test::runTest(f, true)) {
	  std::cout << "PASSED\n";	  
	  count++;
	} else {
	  std::cout << "FAILED\n";	  
	}
    }

    if (count == files.size()) {
      return 42;
    } else {
      return 0;
    }
}
